### PR TITLE
Update AllChem.py

### PR DESCRIPTION
**Pull request by BerendHuisman**
Extended EnumerateLibraryFromReaction to yield not only the product but also reactants which produced that product.

### CODE REVIEW

**Comment by gedeck** [comment]
You change the outside behaviour of a function of RDkit. This can have negative impact in the code of others and as you can see from the Travis runs, it breaks the test for `SimpleEnum.Enumerator` which uses `EnumerateLibraryFromReaction`. In order to avoid this, make the change optional with an explicit opt-in. 
```
def EnumerateLibraryFromReaction(reaction, sidechainSets, returnReactants=False):
...
  for chains in _combiEnumerator(sidechainSets):
    prodSets = reaction.RunReactants(chains)
    for prods in prodSets:
      if returnReactants:
        yield prods, chains
      else:
        yield prods
```
As you return a tuple, you can make use of namedtuples here.
```
from collections import namedtuple
...
  ProductEduct = namedtuple('ProductEduct', 'products,educts')
  for chains in _combiEnumerator(sidechainSets):
    prodSets = reaction.RunReactants(chains)
      if returnReactants:
        yield ProductEduct(prods, chains)
      else:
        yield prods
```
You can then use the result of the function using 'result.products' and 'result.educts' or just use it as a tuple/list.

**Comment by BerendHuisman** [comment]
Thank you for your quick and thourough response. I thought about adding an opt-in but wasn't completly sure if that would be allowed. I will make the changes and make another attempt.
Thank you again for taking the time to review.

**Comment by bp-kelley** [comment]
You may be interested in the enumeration toolkit that was added last year:

http://rdkit.org/Python_Docs/rdkit.Chem.rdChemReactions.EnumerateLibrary-class.html

There is a tutorial that explains how to do what you want:

https://github.com/rdkit/rdkit/tree/master/Code/GraphMol/ChemReactions/tutorial

----
Brian Kelley

File: rdkit/Chem/AllChem.py [hunk 1]
```diff
@@ -260,10 +261,13 @@ def _combiEnumerator(items, depth=0):
       else:
         yield [item]
 
+  ProductEduct = namedtuple('ProductEduct', 'products,educts')
   for chains in _combiEnumerator(sidechainSets):
     prodSets = reaction.RunReactants(chains)
-    for prods in prodSets:
-      yield prods
+      if returnReactants:
+        yield ProductEduct(prods, chains)
```
Comment (gedeck) [inline rdkit/Chem/AllChem.py]: The indentation is messed up. It could be that you used tabs instead of spaces.

**Comment by BerendHuisman** [comment]
Thank you both for the responses. 
@bp-kelley I will definitely take a look at the enumeration toolkit.
@gedeck The indentation should be corrected now. I had one extra space somewhere.
Since I use rdkit on almost a daily basis I wanted to thank you again for helping me and after this experience I am very enthusiastic to contribute more to this amazing project.

File: rdkit/Chem/AllChem.py [hunk 2]
```diff
@@ -264,6 +264,8 @@ def _combiEnumerator(items, depth=0):
   ProductEduct = namedtuple('ProductEduct', 'products,educts')
   for chains in _combiEnumerator(sidechainSets):
     prodSets = reaction.RunReactants(chains)
+    for prods in prodSets:
+      yield prods
       if returnReactants:
         yield ProductEduct(prods, chains)
       else:
```
Comment (gedeck) [inline rdkit/Chem/AllChem.py]: You now yield either prods twice (returnReactants is false) or prods/ProductEduct (returnReactants is True).

File: rdkit/Chem/AllChem.py [hunk 3]
```diff
@@ -260,10 +261,14 @@ def _combiEnumerator(items, depth=0):
...
         yield [item]
 
+  ProductEduct = namedtuple('ProductEduct', 'products,educts')
```
Comment (greglandrum) [inline rdkit/Chem/AllChem.py]: I don't think we should use "educt" here. The rest of the RDKit (at least most of it) uses "reactant". I'd suggest using that in order to remain consistent.

Comment (BerendHuisman) [inline rdkit/Chem/AllChem.py]: Sure, will change immediately

**Comment by greglandrum** [comment]
This looks good to me. If @gedeck doesn't have any objections, I'll go ahead and merge it.

**Comment by BerendHuisman** [comment]
That's great. I'll continue to see what I can contribute to this project. Expect some more pull requests in the future. I look forward to learning from all of you in the future.
One question, when will it be possible to use this functionality through the homebrew installer?

**Comment by greglandrum** [comment]
Thanks for the contribution and welcome to the community!
Once it is merged it will be in master, and will show up in the next release (which I will start in the next week or so).

As a general (disconnected) comment: I'd strongly suggest switching to using anaconda python on the mac instead of using homebrew.

**Comment by BerendHuisman** [comment]
Ok, that sounds great. I'll make the full switch to the anaconda environment then.

**Comment by gedeck** [comment]
Looks good to me.

**Pull request merged**